### PR TITLE
Introduce the index.lifecycle.prefer_ilm setting

### PR DESCRIPTION
**Pull request by andreidan**
This introduces an index setting that'll allow to switch the preference for the system that manages a data stream backing index. This setting defaults to `true` (preferring ILM when both ILM and DLM are applicable for an index).

Note that the prefer_ilm setting will only express a preference, it'll not enforce the presence of a particular management service e.g. if an index has `prefer_ilm: false` and an ILM policy configured, but no DLM configuration, the index will be managed by ILM (despite the preference for DLM)

Note that documentation for the setting will be done in a subsequent PR.
There will also be subsequent work to improve the `GET _data_stream` API to better indicate which system manages each backing index.

Relates to https://github.com/elastic/elasticsearch/issues/93596

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportResumeFollowAction.java [hunk 1]
```diff
@@ -515,7 +515,8 @@ static String[] extractLeaderShardHistoryUUIDs(Map<String, String> ccrIndexMetad
         DataTier.TIER_PREFERENCE_SETTING,
         IndexSettings.BLOOM_FILTER_ID_FIELD_ENABLED_SETTING,
         IndexSettings.LIFECYCLE_ORIGINATION_DATE_SETTING,
-        IndexSettings.LIFECYCLE_PARSE_ORIGINATION_DATE_SETTING
+        IndexSettings.LIFECYCLE_PARSE_ORIGINATION_DATE_SETTING,
```
Comment (andreidan) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportResumeFollowAction.java:518]: Do we all agree this should not be replicated?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportResumeFollowAction.java [hunk 2]
```diff
@@ -515,7 +515,8 @@ static String[] extractLeaderShardHistoryUUIDs(Map<String, String> ccrIndexMetad
...
+        IndexSettings.PREFER_ILM_SETTING
```
Comment (andreidan) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportResumeFollowAction.java:519]: Do we all agree this should not be replicated?

My reasoning is that the leader and follower clusters could have different tiering strategies and use different management systems

Comment (dakrone) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportResumeFollowAction.java:519]: We normally don't replicate settings when they can "double-up" operations. I think we should replicate the setting for now, similar to how we replicate the `index.lifecycle.name` setting

**Review by dakrone — APPROVED** [review]
LGTM, thanks for all the tests for this Andrei, they're great.

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by gmarouli — APPROVED** [review]
Thank you for adding this flag @andreidan . LGTM, I have one minor comment, (apologies if they are there and I missed them) I was thinking it would be nice to a couple of tests also from DLM to ILM, if it's a path we support it should be tests. It can also come from a subsequent PR.

**Comment by andreidan** [comment]
@gmarouli Thanks for the review 

>  was thinking it would be nice to a couple of tests also from DLM to ILM

That's a great shout, I'll add them in a follow up PR if that's ok (this PR is quite big already)

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by andreidan** [comment]
Opened https://github.com/elastic/elasticsearch/issues/95524

@elasticmachine run elasticsearch-ci/part-2

**Pull request merged**